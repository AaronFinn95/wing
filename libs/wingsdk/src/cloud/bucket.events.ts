export interface BucketOnUploadProps {
  /* elided */
}
export interface BucketOnDeleteProps {
  /* elided */
}
export interface BucketOnUpdateProps {
  /* elided */
}
export interface BucketOnEventProps {
  /* elided */
}

export interface BucketEvent {
  key: string;
  type: BucketEventType;
}

export enum BucketEventType {
  PUT = "PUT",
  DELETE = "DELETE",
  UPDATE = "UPDATE",
}

export abstract class BucketEvents implements IBucketEvents {
  private _changeHandlers = new Map<BucketEventType, (key: string) => void>();
  private _onEventFn?: (event: BucketEvent) => void;

  onUpload(fn: (key: string) => void, opts?: BucketOnUploadProps): void {
    if (opts) {
      console.warn("Bucket.onUpload does not support props yet");
    }
    this._changeHandlers.set(BucketEventType.PUT, fn);
  }
  onDelete(fn: (key: string) => void, opts?: BucketOnDeleteProps): void {
    if (opts) {
      console.warn("Bucket.onDelete does not support props yet");
    }
    this._changeHandlers.set(BucketEventType.DELETE, fn);
  }
  onUpdate(fn: (key: string) => void, opts?: BucketOnUpdateProps): void {
    if (opts) {
      console.warn("Bucket.onUpdate does not support props yet");
    }
    this._changeHandlers.set(BucketEventType.UPDATE, fn);
  }
  onEvent(fn: (event: BucketEvent) => void, opts?: BucketOnEventProps): void {
    if (opts) {
      console.warn("Bucket.onEvent does not support props yet");
    }
    this._onEventFn = fn;
  }

  protected async onChange(type: BucketEventType, key: string): Promise<void> {
    try {
      if (this._changeHandlers.has(type)) {
        const fn = this._changeHandlers.get(type);
        try {
          typeof fn == "function" && (await fn(key));
        } catch (error) {
          console.warn(
            `Error while running bucket's "on_${type
              .toString()
              .toLowerCase()}" event: ${error}`
          );
        }
      }
      try {
        typeof this._onEventFn === "function" &&
          (await this._onEventFn({ key, type }));
      } catch (error) {
        console.warn(`Error while running bucket's "on_event" event: ${error}`);
      }
    } catch (error) {
      console.warn(error);
    }
  }
}

export interface IBucketEvents {
  /**
   * Run an inflight whenever a file is uploaded to the bucket.
   */
  onUpload(fn: (key: string) => void, opts?: BucketOnUploadProps): void;

  /**
   * Run an inflight whenever a file is deleted from the bucket.
   */
  onDelete(fn: (key: string) => void, opts?: BucketOnDeleteProps): void;

  /**
   * Run an inflight whenever a file is updated in the bucket.
   */
  onUpdate(fn: (key: string) => void, opts?: BucketOnUpdateProps): void;

  /**
   * Run an inflight whenever a file is uploaded, modified, or deleted from the bucket.
   */
  onEvent(fn: (event: BucketEvent) => void, opts?: BucketOnEventProps): void;
}
