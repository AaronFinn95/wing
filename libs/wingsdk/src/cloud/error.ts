/**
 * Error thrown by a resource while running inflight code.
 */
export class InflightError extends Error {
  constructor(
    /** Error code */
    public readonly code: InflightErrorCode,
    /** Human-readable error message */
    public readonly message: string,
    /** Parent error, if any */
    public readonly cause?: any // Error
  ) {
    super(message);
  }
}

/**
 * Error codes for `InflightError`.
 */
export enum InflightErrorCode {
  /** Indicates that a resource was not found. */
  NOT_FOUND = "NOT_FOUND",
}
