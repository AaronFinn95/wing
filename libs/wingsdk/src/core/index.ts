export * from "./app";
export * from "./attributes";
export * from "./connections";
export * from "./dependency";
export * from "./display";
export * from "./inflight";
export * from "./tree";
