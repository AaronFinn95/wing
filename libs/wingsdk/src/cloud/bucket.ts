import { Construct } from "constructs";
import { Polycons } from "polycons";
import { InflightError, InflightErrorCode } from "./error";
import { Code, Resource } from "../core";

/**
 * Global identifier for `Bucket`.
 */
export const BUCKET_TYPE = "wingsdk.cloud.Bucket";

/**
 * Properties for `Bucket`.
 */
export interface BucketProps {
  /**
   * Whether the bucket's objects should be publicly accessible.
   * @default false
   */
  readonly public?: boolean;
}

/**
 * Functionality shared between all `Bucket` implementations.
 */
export abstract class BucketBase extends Resource {
  public readonly stateful = true;
  constructor(scope: Construct, id: string, props: BucketProps) {
    super(scope, id);

    this.display.title = "Bucket";
    this.display.description = "A cloud object store";

    if (!scope) {
      return;
    }

    props;
  }

  /**
   * Add a file to the bucket that is uploaded when the app is deployed.
   *
   * TODO: In the future this will support uploading any `Blob` type or
   * referencing a file from the local filesystem.
   */
  public abstract addObject(key: string, body: string): void;
}

/**
 * Represents a cloud object store.
 *
 * @inflight `@winglang/sdk.cloud.IBucketClient`
 */
export class Bucket extends BucketBase {
  constructor(scope: Construct, id: string, props: BucketProps = {}) {
    super(null as any, id, props);
    return Polycons.newInstance(BUCKET_TYPE, scope, id, props) as Bucket;
  }

  /** @internal */
  public _toInflight(): Code {
    throw new Error("Method not implemented.");
  }

  public addObject(key: string, body: string): void {
    key;
    body;
    throw new Error("Method not implemented.");
  }
}

/** Interface for delete method inside `Bucket` */
export interface BucketDeleteOptions {
  /**
   * Check failures on the method and retrieve errors if any
   * @Throws if this is `true`, an error is thrown if the file is not found (or any error case).
   * @default false
   */
  readonly mustExist?: boolean;
}

/**
 * Inflight interface for `Bucket`.
 */
export interface IBucketClient {
  /**
   * Put an object in the bucket.
   * @param key Key of the object.
   * @param body Content of the object we want to store into the bucket.
   * @inflight
   */
  put(key: string, body: string): Promise<void>;

  /**
   * Retrieve an object from the bucket.
   * @param key Key of the object.
   * @throws if no object with the given key exists.
   * @returns the object's body.
   * @inflight
   */
  get(key: string): Promise<string>;

  /**
   * Retrieve an object from the bucket, if it exists.
   * @param key Key of the object.
   * @throws if there is any error other than the object not existing.
   * @returns the object's body, if it exists
   * @inflight
   */
  tryGet(key: string): Promise<string | undefined>;

  /**
   * Retrieve existing objects keys from the bucket.
   * @param prefix Limits the response to keys that begin with the specified prefix.
   * @returns a list of keys or an empty array if the bucket is empty.
   * @inflight
   */
  list(prefix?: string): Promise<string[]>;

  /**
   * Delete an existing object using a key from the bucket
   * @param key Key of the object.
   * @param opts Options available for delete an item from a bucket.
   * @inflight
   */
  delete(key: string, opts?: BucketDeleteOptions): Promise<void>;
}

/**
 * Base class for inflight `Bucket` implementations.
 * @internal
 */
export abstract class BucketClientBase implements IBucketClient {
  public abstract put(key: string, body: string): Promise<void>;
  public abstract get(key: string): Promise<string>;
  public async tryGet(key: string): Promise<string | undefined> {
    try {
      return await this.get(key);
    } catch (e) {
      if (
        e instanceof InflightError &&
        e.code === InflightErrorCode.NOT_FOUND
      ) {
        return undefined;
      }
      // We haven't seen this error before, so we rethrow it.
      throw e;
    }
  }
  public abstract list(prefix?: string): Promise<string[]>;
  public abstract delete(
    key: string,
    opts?: BucketDeleteOptions
  ): Promise<void>;
}

/**
 * List of inflight operations available for `Bucket`.
 * @internal
 */
export enum BucketInflightMethods {
  /** `Bucket.put` */
  PUT = "put",
  /** `Bucket.get` */
  GET = "get",
  /** `Bucket.list` */
  LIST = "list",
  /** `Bucket.delete` */
  DELETE = "delete",
}
