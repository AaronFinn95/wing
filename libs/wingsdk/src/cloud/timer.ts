import { Construct } from "constructs";
import { Polycons } from "polycons";
import { Function } from "./function";
import { Code, Resource } from "../core";
import { Duration } from "../std";

/**
 * Global identifier for `Timer`.
 */
export const TIMER_TYPE = "wingsdk.cloud.Timer";

/**
 * Properties for `Timer`.
 */
export interface TimerProps {}

/**
 * Functionality shared between all `Timer` implementations.
 */
export abstract class TimerBase extends Resource {
  public readonly stateful = true;
  constructor(scope: Construct, id: string, props: TimerProps = {}) {
    super(scope, id);

    this.display.title = "Timer";
    this.display.description = "A distributed timer";

    if (!scope) {
      return;
    }

    props;
  }
}

/**
 * Represents a timer.
 *
 * @inflight `@winglang/sdk.cloud.ITimerClient`
 */
export class Timer extends TimerBase {
  constructor(scope: Construct, id: string, props: TimerProps = {}) {
    super(null as any, id, props);
    return Polycons.newInstance(TIMER_TYPE, scope, id, props) as Timer;
  }

  /** @internal */
  public _toInflight(): Code {
    throw new Error("Method not implemented.");
  }
}

/**
 * Inflight interface for `Timer`.
 */
export interface ITimerClient {
  /**
   * Invokes the specified callback after the specified duration.
   * @param after Duration to wait before invoking the callback.
   * @param callback Callback to invoke.
   * @param payload Optional payload to pass to the callback.
   * @inflight
   */
  setTimeout(
    after: Duration,
    callback: Function,
    payload?: string
  ): Promise<void>;
}

/**
 * List of inflight operations available for `Timer`.
 * @internal
 */
export enum TimerInflightMethods {
  /** `Timer.setTimeout` */
  SET_TIMEOUT = "set_timeout",
}
