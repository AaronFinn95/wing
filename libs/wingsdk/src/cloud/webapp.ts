import { mkdirSync, readFileSync, writeFileSync } from "fs";
import { join } from "path";
import { Construct } from "constructs";
import * as esbuild from "esbuild-wasm";
import { Polycons } from "polycons";
import { Code, IApp, Inflight, NodeJsCode, Resource } from "../core";
import { mkdtemp } from "../util";
import { Logger } from "./logger";

export const WEBAPP_TYPE = "wingsdk.cloud.WebApp";

/**
 * Props for `WebApp`.
 */
export interface WebAppProps {}

/**
 * Produces a .zip file that contains a client web application.
 */
export abstract class WebAppBase extends Resource {
  public readonly stateful = false;

  /**
   * The local path to the web application.
   */
  public readonly dist: string;

  constructor(
    scope: Construct,
    id: string,
    client: Inflight,
    props: WebAppProps = {}
  ) {
    super(scope, id);

    if (!scope) {
      this.dist = undefined as any;
      return;
    }

    props;

    const outdir = (this.node.root as IApp).outdir;

    client._bind(this, ["handle"]);

    const archive = join(outdir, "webapps", this.node.addr);
    mkdirSync(archive, { recursive: true });
    writeFileSync(
      join(archive, "index.html"),
      `<!DOCTYPE html>
      <html>
        <head>
          <title>${this.node.path}</title>
        </head>
        <body>
          <h1>${this.node.path}</h1>
          <script src="./client.js"></script>
        </body>
      </html>
    `
    );

    const bundle = readFileSync(bundleInflight(client));
    writeFileSync(
      join(archive, "client.js"),
      `
    const exports = {};

    ${bundle}

    exports.handler()
      .catch(e => console.log(e))
      .then(() => console.log("done"))
    `
    );

    this.dist = archive;
  }

  /**
   * @internal
   */
  public _toInflight(): Code {
    return NodeJsCode.fromInline("bla");
  }
}

/**
 * Represents a webapp.
 *
 * @inflight `@winglang/sdk.cloud.IWebAppClient`
 */
export class WebApp extends WebAppBase {
  constructor(
    scope: Construct,
    id: string,
    client: Inflight,
    props: WebAppProps = {}
  ) {
    super(null as any, id, client, props);
    return Polycons.newInstance(
      WEBAPP_TYPE,
      scope,
      id,
      client,
      props
    ) as WebApp;
  }

  /** @internal */
  public _toInflight(): Code {
    throw new Error("Method not implemented.");
  }
}

function bundleInflight(inflight: Inflight, logger?: Logger) {
  const inflightClient = inflight._toInflight();
  const lines = new Array<string>();

  // create a logger inflight client and attach it to `console.log`.
  // TODO: attach console.error, console.warn, once our logger supports log levels.
  if (logger) {
    const loggerClientCode = logger._toInflight();
    lines.push(`const $logger = ${loggerClientCode.text};`);
    lines.push(`console.log = (...args) => $logger.print(...args);`);
  }

  lines.push("exports.handler = async function(event) {");
  lines.push(`  return await ${inflightClient.text}.handle(event);`);
  lines.push("};");

  const tempdir = mkdtemp();
  const outfile = join(tempdir, "index.js");

  esbuild.buildSync({
    bundle: true,
    stdin: {
      contents: lines.join("\n"),
      resolveDir: tempdir,
      sourcefile: "inflight.js",
    },
    target: "node16",
    platform: "node",
    absWorkingDir: tempdir,
    outfile,
    minify: false,
    external: ["aws-sdk"],
  });

  // the bundled contains line comments with file paths, which are not useful for us, especially
  // since they may contain system-specific paths. sadly, esbuild doesn't have a way to disable
  // this, so we simply filter those out from the bundle.
  const outlines = readFileSync(outfile, "utf-8").split("\n");
  const isNotLineComment = (line: string) => !line.startsWith("//");
  writeFileSync(outfile, outlines.filter(isNotLineComment).join("\n"));

  return outfile;
}
