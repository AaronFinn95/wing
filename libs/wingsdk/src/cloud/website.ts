import { Construct } from "constructs";
import { fqnForType } from "../constants";
import { App, Resource } from "../core";

/**
 * Global identifier for `Website`.
 */
export const WEBSITE_FQN = fqnForType("cloud.Website");

/**
 * Represents a cloud object store.
 *
 * @inflight `@winglang/sdk.cloud.IWebsiteClient`
 */
export abstract class Website extends Resource {
  /**
   * Create a new website.
   * @internal
   */
  public static _newWebsite(
    scope: Construct,
    id: string,
    props: WebsiteProps
  ): Website {
    return App.of(scope).newAbstract(WEBSITE_FQN, scope, id, props);
  }

  public readonly stateful = false;
  private readonly _path: string;
  /**@internal */
  protected _url!: string;
  /** @internal */
  protected _domain?: string;

  constructor(scope: Construct, id: string, props: WebsiteProps) {
    super(scope, id);

    this.display.title = "Website";
    this.display.description = "A static website";

    this._path = props.path;
    this._domain = props.domain;
  }

  /**
   * Local path to the website's static files, relative to the Wing source file.
   * @example "./dist"
   */
  get path(): string {
    return this._path;
  }

  /**
   * The website's url.
   */
  get url(): string {
    return this._url;
  }
}
/**
 * website props
 */
export interface WebsiteProps {
  /**
   * Local path to the website's static files, relative to the Wing source file.
   * @example "./dist"
   */
  readonly path: string;

  /**
   * The website's custom domain name.
   * @example "example.com"
   * @default - a domain is generated by the cloud provider
   */
  readonly domain?: string;
}

/**
/**
 * Inflight methods and members of `cloud.Website`.
 */
export interface IWebsiteClient {}
