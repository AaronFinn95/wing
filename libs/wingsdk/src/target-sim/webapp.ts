import { Construct } from "constructs";
import { WebAppBase, WebAppProps, WEBAPP_TYPE } from "../cloud/webapp";
import { Inflight } from "../core";
import { ISimulatorResource } from "./resource";
import { BaseResourceSchema } from "./schema";
import { WebAppSchema } from "./schema-resources";

export class WebApp extends WebAppBase implements ISimulatorResource {
  constructor(
    scope: Construct,
    id: string,
    client: Inflight,
    props: WebAppProps = {}
  ) {
    super(scope, id, client, props);
  }

  toSimulator(): BaseResourceSchema {
    const schema: WebAppSchema = {
      type: WEBAPP_TYPE,
      path: this.node.path,
      props: {
        dist: this.dist,
      },
      attrs: {} as any,
    };
    return schema;
  }
}
