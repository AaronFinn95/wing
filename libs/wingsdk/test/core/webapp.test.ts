import { WebApp } from "../../src/cloud/webapp";
import { Inflight, NodeJsCode } from "../../src/core";
import { SimApp } from "../../src/testing";

test("foo", async () => {
  const app = new SimApp();

  const js = new Inflight(app, "WebClient", {
    code: NodeJsCode.fromInline(`
      async handle() {
        alert("hello wing in your browser");
      }
    `),
  });

  new WebApp(app, "WebApp", js);

  await app.startSimulator();

  expect(app.snapshot()).toMatchSnapshot();
});
