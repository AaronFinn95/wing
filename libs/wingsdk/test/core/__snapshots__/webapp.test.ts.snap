// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`foo 1`] = `
Object {
  "app.wsim/simulator.json": Object {
    "resources": Array [
      Object {
        "attrs": Object {},
        "path": "root/WingLogger",
        "props": Object {},
        "type": "wingsdk.cloud.Logger",
      },
      Object {
        "attrs": Object {},
        "path": "root/WebApp",
        "props": Object {
          "dist": "/var/folders/vm/cm444l3124j_vh_8rkhz6kqm0000gn/T/wingsdk.hqj04t/webapps/c88eb3bd7278bed1b3d88540951c1b6b5920325e5d",
        },
        "type": "wingsdk.cloud.WebApp",
      },
    ],
    "sdkVersion": "0.0.0",
  },
  "tree.json": Object {
    "tree": Object {
      "children": Object {
        "WebApp": Object {
          "attributes": Object {
            "wing:resource:connections": Array [],
            "wing:resource:stateful": false,
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.130",
          },
          "id": "WebApp",
          "path": "root/WebApp",
        },
        "WebClient": Object {
          "attributes": Object {
            "wing:resource:connections": Array [],
            "wing:resource:stateful": false,
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.130",
          },
          "display": Object {
            "description": "An inflight resource",
            "hidden": true,
            "title": "Inflight",
          },
          "id": "WebClient",
          "path": "root/WebClient",
        },
        "WingLogger": Object {
          "attributes": Object {
            "wing:resource:connections": Array [],
            "wing:resource:stateful": true,
          },
          "constructInfo": Object {
            "fqn": "constructs.Construct",
            "version": "10.0.130",
          },
          "display": Object {
            "description": "A cloud logging facility",
            "hidden": true,
            "title": "Logger",
          },
          "id": "WingLogger",
          "path": "root/WingLogger",
        },
      },
      "constructInfo": Object {
        "fqn": "constructs.Construct",
        "version": "10.0.130",
      },
      "id": "root",
      "path": "root",
    },
    "version": "tree-0.1",
  },
  "webapps/c88eb3bd7278bed1b3d88540951c1b6b5920325e5d/client.js": "
    const exports = {};

    exports.handler = async function(event) {
  return await new (function() {
    return class Handler {
      constructor(clients) {
        for (const [name, client] of Object.entries(clients)) {
          this[name] = client;
        }
      }
      async handle() {
        alert(\\"hello wing in your browser\\");
      }
    };
  }())({}).handle(event);
};


    exports.handler().then(() => console.log(\\"done\\"));
    ",
  "webapps/c88eb3bd7278bed1b3d88540951c1b6b5920325e5d/index.html": "<!DOCTYPE html>
      <html>
        <head>
          <title>root/WebApp</title>
        </head>
        <body>
          <h1>root/WebApp</h1>
          <script src=\\"./client.js\\"></script>
        </body>
      </html>
    ",
}
`;
